examples/step-71/step-71.cc

```diff
@@ -0,0 +1,3765 @@
+    // + \lambda_{e} \ln^{2} \left(\text{det}(\mathbf{F}) \right)
+    // - \frac{1}{2} \mu_{0} \mu_{r} \text{det}(\mathbf{F})
+    //     \left[ \boldsymbol{\mathbb{H}} \cdot \mathbf{C}^{-1} \cdot \boldsymbol{\mathbb{H}} \right]
+    // @f]
+    // with
+    // @f[
+    //  f_{\mu_{e}} \left( \boldsymbol{\mathbb{H}} \right)
+    // = 1 + \left[ \frac{\mu_{e}^{\infty}}{\mu_{e}} - 1 \right]
+    //     \tanh \left( 2 \frac{\boldsymbol{\mathbb{H}} \cdot \boldsymbol{\mathbb{H}}}
+    //       {\left(h_{e}^{sat}\right)^{2}} \right)
+    // @f]
+    // and for which the variable $d = \text{tr}(\mathbf{I})$ ($\mathbf{I}$ being the second-order identity tensor) represents the spatial dimension
+    // and $\mathbf{F}$ is the deformation gradient tensor.
+    // To give some brief background to the various components of $\psi_{0}$, the first
+    // two terms bear a great resemblance to the stored energy density function for a
+    // (hyperelastic) Neohookean material. The only difference between what's used
+    // here and the Neohookean material is the scaling of the elastic shear modulus
+    // by the magnetic field-sensitive saturation function 
+    // $f_{\mu_{e}} \left( \boldsymbol{\mathbb{H}} \right)$ (@citep Pelteret2018a equation 29). This function will,
+    // in effect, cause the material to stiffen in the presence of a strong magnetic
+    // field. As its governed by a sigmoid-type function, the shear modulus will
+    // asymptotically converge on the specified saturation shear modulus.
+    // It can also be shown that the last term in $\psi_{0}$ is the 
+    // stored energy density function for magnetic field (as derived from first principles), scaled by the relative
+    // permeability constant. This definition collectively implies that the material
+    // is linearly magnetized, i.e. the magnetization vector and magnetic field
+    // vector are aligned.
+    //
+    // Since we expect that this class fully describes a single material, we'll
+    // mark it as "final" so that the inheritance tree terminated here.
+    template <int dim, AD::NumberTypes ADTypeCode>
+    class Magnetoelastic_Constitutive_Law_AD final
+      : public Coupled_Magnetomechanical_Constitutive_Law_Base<dim>
+    {
+      // Here we fefine the helper type that we will use in the AD computations for our
+      // scalar energy density function. Note that we expect it to return values of
+      // type double.
+      using ADHelper     = AD::ScalarFunction<dim, ADTypeCode, double>;
+      using ADNumberType = typename ADHelper::ad_type;
+
+    public:
+      Magnetoelastic_Constitutive_Law_AD(
+        const ConstitutiveParameters &constitutive_parameters);
+
+      // Since the public interface to the base class is pure-virtual, here
+      // we'll declare that this class will override all of these base class
+      // methods. 
+      void update_internal_data(const Tensor<1, dim> &         H,
```

Comment:
We don't use `override` alone, we always add `virtual` even if it's not necessary.

Reply:
Ok, I'll go through and add this where necessary.